may-2018-reenabled-opcodes.md

```diff
@@ -0,0 +1,428 @@
+behaviour should fail fast.  
+
+Each op code should be examined for the following risk conditions and mitigating behaviour defined explicitly:
+* Operand byte length mismatch.  Where it would be normally expected that two operands would be of matching byte lengths
+  the resultant behaviour should be defined.
+* Signed integer.  Whether signed integers are permitted operands and whether any special handling is required.
+* Stack size impact.  Both number of elements and total size of elements. 
+* Overflows.  Defined behaviour in the instance that result of the operation exceeds MAX_SCRIPT_ELEMENT_SIZE
+* Empty byte sequence operands.  Whether empty byte sequences should be allowed as a representation of zero.
+* Empty byte sequence output.  Note that an operation that outputs an empty byte sequence has effectively pushed `false` 
+  onto the stack. If this is the last operation in a script or if a conditional operator immediately follows the script 
+  author must consider this possibility. This is currently the case for many existing op codes however so it is 
+  consistent to continue with allowing this behaviour.
+
+## Definitions
+
+* *Stack memory use*. This is the sum of the size of the elements on the stack. It gives an indication of impact on 
+  memory use by the interpreter.
+* *Operand order*. In keeping with convention where multiple operands are specified the top most stack item is the 
+  last operand.  e.g. `x1 x2 OP_CAT` --> `x2` is the top stack item and `x1` is the next from the top.
+* *empty byte sequence*. Throughout this document `OP_0` is used as a convenient representation of an empty byte 
+  sequence.  Whilst it is a push data op code, its effect is to push an empty byte sequence on to the stack.
+
+## Specification
+
+Global conditions apply to all operations. These conditions must be checked by the implementation when 
+it is possible that they will occur:
+* for all e : elements on the stack, `0 <= len(e) <= MAX_SCRIPT_ELEMENT_SIZE`
+* for each operator, the required number of operands are present on the stack when the operand is executed
+
+These unit tests should be included for every operation:
+1. executing the operation with an input element of length greater than `MAX_SCRIPT_ELEMENT_SIZE` will fail
+2. executing the operation with an insufficient number of operands on the stack causes a failure
+
+
+Operand consumption:
+
+In all cases where not explicitly stated otherwise the operand stack elements are consumed by the operation and replaced with the output.
+
+## Splice operations
+
+### OP_CAT
+
+    Opcode (decimal): 126
+    Opcode (hex): 0x7e
+
+Concatenates two operands.
+
+    x1 x2 OP_CAT -> out
+        
+Examples:
+* `Ox11 0x2233 OP_CAT -> 0x112233`
+    
+The operator must fail if:
+* `len(out) > MAX_SCRIPT_ELEMENT_SIZE`. The operation cannot output elements that violate the constraint on the element size.
```

Comment:
For visual clarity and formatting consistency, I would put the sentence starting with "The operation" on a new line and remove the period after the code example. Otherwise, the colon should be removed on the prior line and the whole thing presented as a single line.

Reply:
merged into a single line